### PR TITLE
Database Errors are Hidden when Running GN3

### PR DESCRIPTION
**Pull request by jnduli**
# Background

During set up for GN3, the home page failed loading because of a database error, but the logs didn't explain this  even with `FLASK_DEBUG=1` set. The error I got was:

```
Traceback (most recent call last):
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 2095, in __call__
    return self.wsgi_app(environ, start_response)
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 2080, in wsgi_app
    response = self.handle_exception(e)
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 2077, in wsgi_app
    response = self.full_dispatch_request()
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 1526, in full_dispatch_request
    return self.finalize_request(rv)
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 1545, in finalize_request
    response = self.make_response(rv)
  File "/gnu/store/1hqil50cpaq32rhl4q0p1inr25png6kx-python-flask-2.1.1/lib/python3.10/site-packages/flask/app.py", line 1701, in make_response
    raise TypeError(
TypeError: The view function for 'menu.generate_json' did not return a valid response. The function either returned None or ended without a return statement.

```

# Solution

The `db_utils.database_connection` context manager was masking away these errors. To fix this, I've added support for logging an error here. This shouldn't change any functionality when using the module. 

# Tested

Running the same task now results in a clearer message:
```
DB error encountered
Traceback (most recent call last):
  File "/home/rookie/gn/gn3/gn3/db_utils.py", line 39, in database_connection
    yield connection
.
.
    db.query(q)
  File "/home/rookie/.guix-extra-profiles/gn3/lib/python3.10/site-packages/MySQLdb/connections.py", line 259, in query
    _mysql.connection.query(self, query)
MySQLdb._exceptions.OperationalError: (1054, "Unknown column 'Family' in 'field list'")

```

### CODE REVIEW

File: gn3/db_utils.py [hunk 1]
```diff
@@ -24,8 +25,10 @@ def cursor(self, *args, **kwargs) -> Any:
 
 
 @contextlib.contextmanager
-def database_connection(sql_uri) -> Iterator[Connection]:
+def database_connection(sql_uri: str, logger: Optional[logging.Logger] = None) -> Iterator[Connection]:
     """Connect to MySQL database."""
+    if logger is None:
```
Comment (fredmanglis) [inline gn3/db_utils.py]: Hey @jnduli 

*suggestion*: Why not move the `logger = logging.getLogger(...)` out of this function, to the module's top level, then pass it as the default logger. That way you can get rid of the `if logger is None` check and the logger is also available to any other function that might need logging in the module.

File: gn3/db_utils.py [hunk 2]
```diff
@@ -35,6 +38,7 @@ def database_connection(sql_uri) -> Iterator[Connection]:
     try:
         yield connection
     except mdb.Error as _mbde:
+        logger.error("DB error encountered", exc_info=1)
```
Comment (fredmanglis) [inline gn3/db_utils.py:42]: **TIL**: I missed `exc_info` as a usable argument. Nice! I have been using `traceback.format_exc` and `traceback.format_exception` when I wanted the exception's stack trace information. I'll try this one out too.

Thanks.

File: gn3/db_utils.py [hunk 3]
```diff
@@ -24,8 +25,10 @@ def cursor(self, *args, **kwargs) -> Any:
 
 
 @contextlib.contextmanager
-def database_connection(sql_uri) -> Iterator[Connection]:
+def database_connection(sql_uri: str, logger: Optional[logging.Logger] = None) -> Iterator[Connection]:
```
Comment (fredmanglis) [inline gn3/db_utils.py]: Well, if you follow my suggestion to move the creation of the logger to the module top level, you'll have to get rid of the use of `Optional` here, and line 4 above - I think that's a good (well, at the very least, acceptable) thing.

**Review by fredmanglis — CHANGES_REQUESTED** [review]
Have a look at the reviews I have provided, make the change and you can merge it. Thanks.

**Review by fredmanglis — APPROVED** [review]
Hi @jnduli 

This looks good. Once you have made the changes I requested in the other commit, go right ahead and merge this.

File: gn3/db_utils.py (hunk 1 above)
Comment (jnduli) [reply gn3/db_utils.py]: Thanks. This makes the implementation much cleaner. I've pushed the suggestion.

File: gn3/db_utils.py (hunk 3 above)
Comment (jnduli) [reply gn3/db_utils.py]: You're right. This is a good thing. I've also dropped the `Optional` import.

**Review by BonfaceKilz — APPROVED** [review]
LGTM!

**Pull request merged**